modules/apps/document-library/document-library-item-selector-web/src/main/java/com/liferay/document/library/item/selector/web/internal/criterion/DLItemSelectorCriterionCreationMenuFilterProviderUtil.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.document.library.item.selector.web.internal.criterion;
+
+import com.liferay.document.library.item.selector.criterion.DLItemSelectorCriterionCreationMenuFilterProvider;
+import com.liferay.item.selector.ItemSelectorCriterion;
+import com.liferay.osgi.service.tracker.collections.map.PropertyServiceReferenceMapper;
+import com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMap;
+import com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMapFactory;
+import com.liferay.petra.reflect.GenericUtil;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Adolfo Pérez
+ */
+@Component(immediate = true, service = {})
+public class DLItemSelectorCriterionCreationMenuFilterProviderUtil {
+
+	public static Set<String> getCreationMenuFilter(
+		ItemSelectorCriterion itemSelectorCriterion) {
+
+		Class<? extends ItemSelectorCriterion> clazz =
+			itemSelectorCriterion.getClass();
+
+		List<DLItemSelectorCriterionCreationMenuFilterProvider>
+			dlItemSelectorCriterionCreationMenuFilterProviders =
+				_serviceTrackerMap.getService(clazz.getName());
+
+		if (dlItemSelectorCriterionCreationMenuFilterProviders == null) {
+			return Collections.emptySet();
+		}
+
+		Stream<DLItemSelectorCriterionCreationMenuFilterProvider> stream =
+			dlItemSelectorCriterionCreationMenuFilterProviders.stream();
+
+		return stream.map(
+			DLItemSelectorCriterionCreationMenuFilterProvider::
+				getAllowedCreationMenuUIItemKeys
+		).collect(
+			HashSet::new, Set::addAll, Set::addAll
+		);
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openMultiValueMap(
+			bundleContext,
+			DLItemSelectorCriterionCreationMenuFilterProvider.class, null,
+			(serviceReference, emitter) -> {
+				Object modelClassName = serviceReference.getProperty(
+					"model.class.name");
```

Comment:
Hi @adolfopa,

When I read `model.class.name` I always expect the property to contain model class names (like FileEntry or JournalArticle). In this case it will be an item selector criterion class, which might be confusing...

What do you think?

Reply:
The rest looks 👍 .

Reply:
As I understand it, the "model.class.name" property defines a scope for the component, so I think it is reasonable to use it here as well -- even if criteria are not service builder models.

Are you ok with leaving this as is? I'd prefer not to introduce a new property just for this case. I say we let Brian decide :-)